zmq/green/core.py

```diff
@@ -95,30 +95,36 @@ def __setup_events(self):
         self.__writable = AsyncResult()
         self.__readable.set()
         self.__writable.set()
-        
+
+        fd = self.getsockopt(zmq.FD)
         try:
-            self._state_event = get_hub().loop.io(self.getsockopt(zmq.FD), 1) # read state watcher
-            self._state_event.start(self.__state_changed)
+            # read state watcher
+            self._state_event = get_hub().loop.io(fd, 1)
+            self._state_event.start(self.__readable_detected)
         except AttributeError:
             # for gevent<1.0 compatibility
             from gevent.core import read_event
-            self._state_event = read_event(self.getsockopt(zmq.FD), self.__state_changed, persist=True)
+            self._state_event = read_event(fd, self.__readable_detected, persist=True)
+
+    def __readable_detected(self):
+        # NOTE: DO NOT MAKE BLOCKING HERE
+        if self.closed:
+            self.__cleanup_events()
+            return
+        self.__readable.set()
 
     def __state_changed(self, event=None, _evtype=None):
         if self.closed:
             self.__cleanup_events()
             return
-        try:
-            # avoid triggering __state_changed from inside __state_changed
-            events = super(_Socket, self).getsockopt(zmq.EVENTS)
-        except zmq.ZMQError as exc:
-            self.__writable.set_exception(exc)
-            self.__readable.set_exception(exc)
-        else:
-            if events & zmq.POLLOUT:
-                self.__writable.set()
-            if events & zmq.POLLIN:
-                self.__readable.set()
+        # getsockopt(ZMQ_EVENTS) at here can cause SIGABRT
+        # https://github.com/zeromq/libzmq/issues/2942
+        fd = self.getsockopt(zmq.FD)
+        readable, writable, __ = select([fd], [fd], [], timeout=0)
```

Comment:
This is not correct. The zmq FD is only ever readable, never writable. The FD being readable only tells you that there *might* be read or write events waiting to process. The readable and writable events can only ever be set by interrogating ZMQ_EVENTS.

Reply:
Oh, okay.  I didn't know about the difference between `ZMQ_EVENTS` and `select()`.

It seems to be the only way to fix the crash that we fix the original issue from libzmq.